compiler/rustc_lint/src/non_fmt_panic.rs

```diff
@@ -52,13 +52,25 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
                 if Some(def_id) == cx.tcx.lang_items().begin_panic_fn()
                     || Some(def_id) == cx.tcx.lang_items().panic_fn()
                     || Some(def_id) == cx.tcx.lang_items().panic_str()
+                    || Some(def_id) == cx.tcx.lang_items().unreachable_display()
                 {
                     if let Some(id) = f.span.ctxt().outer_expn_data().macro_def_id {
+                        let diagnostic_item = cx.tcx.get_diagnostic_name(id);
                         if matches!(
-                            cx.tcx.get_diagnostic_name(id),
+                            diagnostic_item,
                             Some(sym::core_panic_2015_macro | sym::std_panic_2015_macro)
                         ) {
-                            check_panic(cx, f, arg);
+                            check_panic_or_unreachable(cx, f, arg);
+                        } else if matches!(diagnostic_item, Some(sym::unreachable_2015_macro)) {
```

Comment:
It's probably better to do the removal of the `AddrOf` operator based on the function that was called (`unreachable_display()`), rather than the macro (`unreachable!()`), since `unreachable!()` can also expand to `panic_fn` where this shouldn't happen.